rest/doc_api.go

```diff
@@ -356,6 +356,58 @@ func (h *handler) handlePutAttachment() error {
 	return nil
 }
 
+func (h *handler) handleDeleteAttachment() error {
+	if h.isReadOnlyGuest() {
+		return base.HTTPErrorf(http.StatusForbidden, auth.GuestUserReadOnly)
+	}
+
+	docid := h.PathVar("docid")
+	attachmentName := h.PathVar("attach")
+	revid := h.getQuery("rev")
+	if revid == "" {
+		var err error
+		revid, err = h.getEtag("If-Match")
+		if err != nil {
+			return err
+		}
+	}
+
+	body, err := h.db.Get1xRevBody(h.ctx(), docid, revid, false, nil)
+	if err != nil {
+		if base.IsDocNotFoundError(err) {
+			// Need to return an error if a document is not found
+			return base.HTTPErrorf(http.StatusNotFound, "Document specified is not found")
+		} else if err != nil {
+			return err
+		}
+	} else if body != nil {
+		if revid == "" {
+			// If a revid is not specified and an active revision was found,
+			// return a conflict now, rather than letting db.Put do it further down...
+			return base.HTTPErrorf(http.StatusConflict, "Cannot modify attachments without a specific rev ID")
+		}
+	}
+
+	// get document attachments and check if attachment exists
+	attachments := db.GetBodyAttachments(body)
+	if _, ok := attachments[attachmentName]; !ok {
+		return base.HTTPErrorf(http.StatusNotFound, "Attachment %s is not found", attachmentName)
+	}
+	// delete specified attachment from the map
```

Comment:
There's quite a bit of shared code between this and PUT attachment - it looks like the next two lines are the only thing that differs.  Did you evaluate whether it made sense to move some of the shared functionality to a shared function?  (it might be valid to keep it separate, just want to understand the reasoning)

Reply:
I think you raise a good point. My reasoning for not having a shared helper function was that some of the shared code areas have some small differences. For example the code at the top is not pulling content type header as its not needed. The code after fetching the body, when getting document not found, checks if the error is relating to incorrect revid or not and returns the appropriate status code (an issue I spotted after your comment about adding test case for specifying incorrect revid for delete endpoint). Then the code to check if the attachment exists on the document or not. 
I am probably wrong with my thinking on that though, do you think I should separate the lines 360 to 393 into a separate shared function and get a way of executing the slightly different logic based off if the function is being called from PUT or DELETE endpoints?

Reply:
I think there are enough differences there that it's worth just leaving them without shared code - thanks for the additional information.